### PR TITLE
NAM-3269: Fix allauth state

### PR DESCRIPTION
**Pull request by annamalaissh**

### CODE REVIEW

File: allauth/socialaccount/providers/oauth2/client.py [hunk 1]
```diff
@@ -29,14 +29,16 @@ def __init__(self, request, consumer_key, consumer_secret,
...
-            'redirect_uri': self.callback_url,
             'scope': self.scope,
             'response_type': 'code'
         }
         if self.state:
             params['state'] = self.state
         params.update(extra_params)
-        return '%s?%s' % (authorization_url, urlencode(params))
+        redirect_url = '%s?%s' % (authorization_url, urlencode(params))
+        redirect_url += '&redirect_uri=%s' % self.callback_url
```
Comment (annamalaissh) [inline allauth/socialaccount/providers/oauth2/client.py]: redirect_uri which is the callback url will have the encoded_state as a get parameter. URL Encoding the callback url results fb throwing error that the redirect url is not in apps settings. So redirect uri is not url encoded

File: allauth/socialaccount/providers/oauth2/views.py [hunk 2]
```diff
@@ -84,6 +85,9 @@ def dispatch(self, request):
         auth_url = self.adapter.authorize_url
         auth_params = provider.get_auth_params(request, action)
         client.state = SocialLogin.stash_state(request)
+        state = SocialLogin.state_from_request(request)
+        encoded_state = pickle.dumps(state).encode('base64', 'strict')
+        client.callback_url += '?encoded_state=%s' % encoded_state
```
Comment (vladimir-socialsweethearts) [inline allauth/socialaccount/providers/oauth2/views.py]: we include state 2 times here?

1. here as ?encoded_state
2. in client.get_redirect_url as ?state=...&client_id=...&redirect_uri=...&scope=...&response_type=...

why do we need to include it twice?
beside that the url will be broken with 2 question marks.

File: allauth/socialaccount/providers/oauth2/views.py [hunk 3]
```diff
@@ -119,10 +127,16 @@ def dispatch(self, request):
                                                 response=access_token)
             login.token = token
             if self.adapter.supports_state:
-                login.state = SocialLogin \
-                    .verify_and_unstash_state(
-                        request,
-                        get_request_param(request, 'state'))
+                try:
```
Comment (vladimir-socialsweethearts) [inline allauth/socialaccount/providers/oauth2/views.py]: we check the state twice here. why do we do this?
state and encoded_state

Comment (vladimir-socialsweethearts) [reply allauth/socialaccount/providers/oauth2/views.py]: It is a verifier (random string) and not the whole state. It makes sense to pass the state along with the callback_url

**Pull request merged**